File: sol/solution/0001-0100/0015/Solution.ex
defmodule Solution do
  def three_sum(nums) do
    nums =
      nums
      |> Enum.sort()
      |> List.to_tuple()
    n = tuple_size(nums)
    0..(n - 3)
    |> Enum.reduce([], fn i, acc ->
      if i > 0 and elem(nums, i) == elem(nums, i - 1) do
        acc
      else
        two_pointer(nums, i, i + 1, n - 1, acc)
      end
    end)
    |> Enum.reverse()
  end
  defp two_pointer(nums, i, left, right, acc) when left < right do
    a = elem(nums, i)
    b = elem(nums, left)
    c = elem(nums, right)
    sum = a + b + c
    cond do
      sum < 0 ->
        two_pointer(nums, i, left + 1, right, acc)
      sum > 0 ->
        two_pointer(nums, i, left, right - 1, acc)
      true ->
        left = move_left(nums, left, right, b)
        right = move_right(nums, left, right, c)
        two_pointer(
          nums,
          i,
          left,
          right,
          [[a, b, c] | acc]
        )
    end
  end
  defp two_pointer(_, _, _, _, acc), do: acc
  defp move_left(nums, left, right, value) do
    cond do
      left < right and elem(nums, left) == value ->
        move_left(nums, left + 1, right, value)
      true ->
        left
    end
  end
  defp move_right(nums, left, right, value) do
    cond do
      left < right and elem(nums, right) == value ->
        move_right(nums, left, right - 1, value)
      true ->
        right
    end
  end
end

File: sol/solution/0001-0100/0015/Solution.scala
object Solution {
  def threeSum(nums: Array[Int]): List[List[Int]] = {
    scala.util.Sorting.quickSort(nums)
    var result = List[List[Int]]()
    for (i <- 0 until nums.length - 2) {
      if (i == 0 || nums(i) != nums(i - 1)) {
        var left = i + 1
        var right = nums.length - 1
        while (left < right) {
          val sum = nums(i) + nums(left) + nums(right)
          if (sum == 0) {
            result ::= List(nums(i), nums(left), nums(right))
            while (left < right && nums(left) == nums(left + 1)) left += 1
            while (left < right && nums(right) == nums(right - 1)) right -= 1
            left += 1
            right -= 1
          } else if (sum < 0) {
            left += 1
          } else {
            right -= 1
          }
        }
      }
    }
    result.reverse
  }
}